pinot-core/src/main/java/org/apache/pinot/core/query/optimizer/filter/NumericalFilterOptimizer.java

```diff
@@ -346,6 +365,156 @@ private static Expression rewriteRangeExpression(Expression range, FilterKind ki
     return range;
   }
 
+  /**
+   * Rewrite expressions of the form "column BETWEEN lower AND upper" to ensure that lower and upper bounds are the same
+   * datatype as the column (or can be cast to the same datatype in the server).
+   */
+  private static Expression rewriteBetweenExpression(Expression between, DataType dataType) {
+    List<Expression> operands = between.getFunctionCall().getOperands();
+    Expression lower = operands.get(1);
+    Expression upper = operands.get(2);
+
+    if (lower.isSetLiteral()) {
+      switch (lower.getLiteral().getSetField()) {
+        case LONG_VALUE: {
+          long actual = lower.getLiteral().getLongValue();
+          // Other data types can be converted on the server side.
```

Comment:
Seems like this is a reduced version of the existing `rewriteRangeExpression()`. To keep the behavior consistent, are we able to extract out the logic of rewriting lower bound and upper bound, then use the common logic for both `rewriteRangeExpression()` and `rewriteBetweenExpression()`?

Reply:
I did consider consolidating the two but the major difference is that in `rewriteRangeExpression` we're rewriting the operators too (and avoiding manually modifying the literal value other than the cast) which we don't want to do for `BETWEEN` (as discussed in https://github.com/apache/pinot/pull/14111/). The other alternative would be to change `rewriteRangeExpression` to avoid rewriting the operators but I thought it'd be better not to change the behavior of the existing optimizer?

Reply:
This part I still don't follow. Do you mean we should not rewrite `BETWEEN` the same way as other range filters, or is it too complicated? As long as we computed lower and upper bound, we should be able to assemble it back to a `BETWEEN`.
It is also fine to do it separately

Reply:
> Do you mean we should not rewrite BETWEEN the same way as other range filters

Yeah, basically this. For instance, taking `intCol >= 2.5` as an example. `2.5` is cast to `2` (int), and then the `>=` is rewritten to `>` because `actual - converted > 0` resulting in `intCol > 2`. For `BETWEEN`, we want to instead rewrite `intCol BETWEEN 2.5 AND y` to `intCol BETWEEN 3 AND y`. We could change the logic for regular range filter to rewrite `intCol >= 2.5` to `intCol > 3` instead to match the `BETWEEN` rewrite logic - is that what you're suggesting? There are some other differences too though. For instance, `floatCol < longLiteral` can be rewritten to `floatCol <= castedFloatLiteral` depending on the comparison between `longLiteral` and `castedFloatLiteral`. We can't do the same for `BETWEEN` though, and we simply skip any conversion in these cases, allowing the server to do the cast. Given these differences, it seemed better overall to keep these rewrites separate, what do you think?

Reply:
Thanks for the detailed explanation. Ideally we want the behavior to be consistent across all range filters. Given the difficulties and the scope of this PR, let's add a TODO to revisit in the future